### ISSUE TITLE
32-bit VS2019 build stalls when exiting Win 3.11

### ISSUE THREAD

**Issue by emendelson** [joncampbell123/dosbox-x#2310]
I've been puzzled by a problem with my system for running Windows 3.11 with DOSBox-X, downloadable here:

http://www.columbia.edu/~em36/Win31DOSBox.html

It uses a slightly customized 32-bit build of DOSBox-X, and I've found that it often doesn't exit properly when I exit Windows. Some of the Win311 screen remains visible, then a black window remains visible, and sometimes needs to be shut down from the close button in the title bar. I thought this was a problem with my build, but instead, the same problem occurs when I replace my build with the standard 32-bit SDL1 current version, downloadable from the Releases page.

The problem doesn't occur with 64-bit builds (both the standard build and my own custom code). Something seems to go wrong in the 32-bit VS2019 builds, but I can't imagine what it might be. I don't expect a fix, but I thought this might be worth reporting in case it's helpful.

**Comment by josephillips85** [comment]
+1 on mac version

**Comment by Wengier** [comment]
@emendelson Have you tried 32-bit VS2019 SDL2 build and 32-bit MinGW SDL1/SDL2 builds? Do they work in this case?

I did try to replace DOSBox-X.exe in the DOSBox folder with a 32-bit VS2019 SDL1 build and ran it, but out of numerous (50+) tries the issue may have only occurred once (which I am also not sure if it is caused by something else this time), and I was never able to produce it again since then. It is possible that some operation(s) need to be done before exiting Win3.11 (instead of just launching and quickly exiting) for the issue to occur, or it may be related to the actual copy of Win3.x (since this is not included in the system itself).

**Comment by joncampbell123** [comment]
@emendelson Another thought is that it may be related to the autoa20fix option that attempts to work around "Packed file is corrupt" errors by switching off the A20 gate.

**Comment by emendelson** [comment]
@Wengier Yes, further testing shows that this problem occurs almost always if I do something in Win311 before exiting. I've been opening Write and pasting from the clipboard, then closing Write and exiting Win311 - and the problem occurs every time (or almost every time). It typically does not occur if I merely open Win311, then press Alt-F4, and exit Win311.

The problem also occurs in the 32-bit MinGW SDL1 builds. The problem does NOT occur in 32-bit SDL2 builds - but, no matter how much I experiment with settings, I cannot get Clipboard Paste to work correctly with Write in SDL2 builds. (Sometimes carriage returns seem to be entered, but no text.) As far as I can see, the only version that lets Win311 exit cleanly AND use Clipboard Paste is the 64-bit SDL1 build. 

@joncampbell123 - I'll be interested in knowing if your comment points to a solution!

**Comment by joncampbell123** [comment]
@emendelson No, that's exactly what I was referring to. Might be something else.

**Comment by Wengier** [comment]
@emendelson You are right that clipboard pasting did not work for Win3.x apps in Windows SDL2 builds because it used BIOS function instead of the method in the Windows SDL1 builds. I have already made Windows SDL2 builds to use the same method as Windows SDL1 builds so that clipboard pasting should work in all Windows builds. You can still use the BIOS function for pasting by selecting the menu option "Use BIOS function for keyboard pasting" in Windows builds.

Furthermore, even if I opened Write and pasted from the clipboard, then closed Write and exited Win311, it is still not easy to reproduce the original problem you reported on my end. So it may be related to the actual copy of Win3.x or so.

**Comment by emendelson** [comment]
@Weingier - The copy of Win311 that I use for testing is the US English version of Windows 3.11 (not Windows for Workgroups, for faster installation). I use a version of the InstallWin31DOSBox.exe program in my download that automatically installs Win311 from the contents of the US English install disks, to make sure that I am testing with the same system every time I install.

**Comment by josephillips85** [comment]
I’m confuse here
3.11 is windows for workgroup and 3.1 is not for workgroup

I’m having this issue on windows 3.11 for workgroup

I got a idea if you are using 3.1 not for workgroup try adding win /s this
will not use the 386 mode just to see if
The issue persist

**Comment by emendelson** [comment]
@josephillips85 Sorry for the confusion. Every post in this thread is about Windows 3.11 or Windows for WorkGroups 3.11. Any reference to 3.1 (not 3.11) was a typing error.
